geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryObserverHolder.java

```diff
@@ -54,26 +54,26 @@
   /**
    * Set the given observer to be notified of query events. Returns the current observer.
    */
-  public static QueryObserver setInstance(QueryObserver observer) {
+  public static synchronized QueryObserver setInstance(QueryObserver observer) {
```

Comment:
Rather than using `synchronized`, you'd be better off making `_instance` a `private static final AtomicReference<QueryObserver>`. The use of `synchronized` will cause at least some JVM implementations to pin every thread hitting it to the same CPU.

Using atomics is generally always a better approach if you can use it.